db/middleware.go

```diff
@@ -87,6 +91,43 @@ func SearchQueryTransformMiddleware() func(echo.HandlerFunc) echo.HandlerFunc {
 	}
 }
 
+func CleanupAgentResources() func(echo.HandlerFunc) echo.HandlerFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+
+			if !strings.Contains(c.Request().URL.Path, "db/agents") ||
+				c.Request().Method != http.MethodPatch {
+				return next(c)
+			}
+
+			// path: /api/db/agents?id=eq.018ef600-0cb0-9a27-d293-8399b9705fbd
+			agentID := strings.Replace(c.QueryParam("id"), "eq.", "", 1)
+
+			var reqBody struct {
+				Cleanup   bool   `json:"cleanup"`
+				DeletedAt string `json:"deleted_at"`
+			}
+			if err := c.Bind(&reqBody); err != nil {
+				return api.WriteError(c, api.Errorf(api.EINVALID, "invalid request: %v", err))
+			}
+
+			if !reqBody.Cleanup || reqBody.DeletedAt != "" {
+				return next(c)
+			}
+
+			if _, err := uuid.Parse(agentID); err != nil {
+				return api.WriteError(c, api.Errorf(api.EINVALID, "unable to parse agent_id (%s) as uuid: %v", agentID, err))
+			}
+
+			ctx := c.Request().Context().(context.Context)
+			if err := cleanupAgentResources(ctx, agentID); err != nil {
```

Comment:
@yashmehrotra couldn't we make this a postgresql func & frontend simply calls that via postgREST?

Also not sure why we're not using a regular handler and going with a middleware.

Reply:
> Also not sure why we're not using a regular handler and going with a middleware.
We have to "intercept" db calls via postgrest so that their normal functionality retains intact

> couldn't we make this a postgresql func & frontend simply calls that via postgREST?

I think thats a better idea, a func would mark everything deleted and it all would happen in one transaction

@moshloop What do you think ?

Reply:
Cleanup agent can be a db function, but I think there should be a trigger on on agents that fires a agent.cleanup async event that calls that function or runs the queries. 

The UI can also call that function, but we shouldn't rely on it